src/c_string.rs

```diff
@@ -0,0 +1,594 @@
+                //
+                // SAFETY: We append a missing nul terminator right below.
+                unsafe {
+                    self.extend_from_bytes_unchecked(bytes).unwrap();
+                    self.inner.push_unchecked(0);
+                };
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Removes the nul byte terminator from the inner buffer.
+    ///
+    /// # Safety
+    ///
+    /// Callers must ensure to add the nul terminator back after this function is called.
+    #[inline]
+    unsafe fn pop_terminator(&mut self) {
+        debug_assert_eq!(self.inner.last(), Some(&0));
+
+        // SAFETY: We always have the nul terminator at the end.
+        unsafe { self.inner.pop_unchecked() };
+    }
+
+    /// Removes the existing nul terminator and then extends `self` with the given bytes.
+    ///
+    /// # Safety
+    ///
+    /// If `additional` is not nul-terminated, the [`CString`] is left non nul-terminated, which is
+    /// an invalid state. Caller must ensure that either `additional` has a terminating nul byte
+    /// or ensure to append a trailing nul terminator.
+    unsafe fn extend_from_bytes_unchecked(
+        &mut self,
+        additional: &[u8],
+    ) -> Result<(), CapacityError> {
+        self.pop_terminator();
+
+        self.inner.extend_from_slice(additional)
+    }
+
+    /// Returns the underlying byte slice including the trailing nul terminator.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut c_string = CString::<5>::new();
+    /// c_string.extend_from_bytes(b"abc").unwrap();
+    ///
+    /// assert_eq!(c_string.as_bytes_with_nul(), b"abc\0");
+    /// ```
+    #[inline]
+    pub fn as_bytes_with_nul(&self) -> &[u8] {
+        &self.inner
+    }
+
+    /// Returns the underlying byte slice excluding the trailing nul terminator.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut c_string = CString::<5>::new();
+    /// c_string.extend_from_bytes(b"abc").unwrap();
+    ///
+    /// assert_eq!(c_string.as_bytes(), b"abc");
+    /// ```
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        debug_assert_eq!(self.inner.last().copied(), Some(0));
+
+        &self.inner[..self.inner.len() - 1]
+    }
+}
+
+impl<const N: usize> AsRef<CStr> for CString<N> {
+    #[inline]
+    fn as_ref(&self) -> &CStr {
+        self.as_c_str()
+    }
+}
+
+impl<const N: usize> Borrow<CStr> for CString<N> {
+    #[inline]
+    fn borrow(&self) -> &CStr {
+        self.as_c_str()
+    }
+}
+
+impl<const N: usize> Default for CString<N> {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<const N: usize> Deref for CString<N> {
+    type Target = CStr;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_c_str()
+    }
+}
+
+impl<const N: usize, T: AsRef<CStr>> PartialEq<T> for CString<N> {
```

Comment:
This doesn't exist in `std`. Only `PartialEq<M> for CString<N>` should be enough.